From 1a4b04749722ba4b8773dc847bb1074f9bc3ab95 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Thu, 9 May 2019 17:11:32 +0200
Subject: [PATCH 156/345] MIPS: OCTEON: Save/Restore wider multiply registers
 in OCTEON III CPUs

The wide multiplier is twice as wide, so we need to save twice as much
state.  Detect the multiplier type (CPU type) at start up and install
model specific handlers.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Leonid Rosenboim <lrosenboim@caviumnetworks.com>
Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
---
 arch/mips/include/asm/octeon/octeon.h | 13 +++++++++++++
 arch/mips/kernel/octeon_switch.S      | 20 +++++++-------------
 2 files changed, 20 insertions(+), 13 deletions(-)

diff --git a/arch/mips/include/asm/octeon/octeon.h b/arch/mips/include/asm/octeon/octeon.h
index 16080c56ad61..f57df81deba1 100644
--- a/arch/mips/include/asm/octeon/octeon.h
+++ b/arch/mips/include/asm/octeon/octeon.h
@@ -479,5 +479,18 @@ int unregister_co_cache_error_notifier(struct notifier_block *nb);
 
 extern unsigned long long cache_err_dcache[NR_CPUS];
 
+/* Octeon multiplier save/restore routines from octeon_switch.S */
+void octeon_mult_save(void);
+void octeon_mult_restore(void);
+void octeon_mult_save_end(void);
+void octeon_mult_restore_end(void);
+void octeon_mult_save3(void);
+void octeon_mult_save3_end(void);
+void octeon_mult_save2(void);
+void octeon_mult_save2_end(void);
+void octeon_mult_restore3(void);
+void octeon_mult_restore3_end(void);
+void octeon_mult_restore2(void);
+void octeon_mult_restore2_end(void);
 
 #endif /* __ASM_OCTEON_OCTEON_H */
diff --git a/arch/mips/kernel/octeon_switch.S b/arch/mips/kernel/octeon_switch.S
index e3f4037b2114..0866c3fc5575 100644
--- a/arch/mips/kernel/octeon_switch.S
+++ b/arch/mips/kernel/octeon_switch.S
@@ -502,6 +502,7 @@ octeon_mult_save3_end:
 	.p2align 7
 	.set push
 	.set noreorder
+	.set arch=octeon3
 	LEAF(octeon_mult_restore)
 	jr	ra
 	 nop
@@ -534,28 +535,21 @@ octeon_mult_restore2_end:
 	ld	$13, PT_MPL+(3*8)(sp)	/* read MPL3 */
 	ld	$10, PT_MPL+(1*8)(sp)	/* read MPL1 */
 	ld	$11, PT_MPL+(4*8)(sp)	/* read MPL4 */
-	.word	0x718d0008
-	/* mtm0	$12, $13		   restore MPL0 and MPL3 */
+	mtm0	$12, $13		/* restore MPL0 and MPL3 */
 	ld	$12, PT_MPL+(2*8)(sp)	/* read MPL2 */
-	.word	0x714b000c
-	/* mtm1	$10, $11		   restore MPL1 and MPL4 */
+	mtm1	$10, $11		/* restore MPL1 and MPL4 */
 	ld	$13, PT_MPL+(5*8)(sp)	/* read MPL5 */
 	ld	$10, PT_MTP+(0*8)(sp)	/* read P0 */
 	ld	$11, PT_MTP+(3*8)(sp)	/* read P3 */
-	.word	0x718d000d
-	/* mtm2	$12, $13		   restore MPL2 and MPL5 */
+	mtm2	$12, $13		/* restore MPL2 and MPL5 */
 	ld	$12, PT_MTP+(1*8)(sp)	/* read P1 */
-	.word	0x714b0009
-	/* mtp0	$10, $11		   restore P0 and P3 */
+	mtp0	$10, $11		/* restore P0 and P3 */
 	ld	$13, PT_MTP+(4*8)(sp)	/* read P4 */
 	ld	$10, PT_MTP+(2*8)(sp)	/* read P2 */
 	ld	$11, PT_MTP+(5*8)(sp)	/* read P5 */
-	.word	0x718d000a
-	/* mtp1	$12, $13		   restore P1 and P4 */
+	mtp1	$12, $13		/* restore P1 and P4 */
 	jr	ra
-	.word	0x714b000b
-	/* mtp2	$10, $11		   restore P2 and P5 */
-
+	  mtp2	$10, $11		/* restore P2 and P5 */
 octeon_mult_restore3_end:
 	EXPORT(octeon_mult_restore3_end)
 	END(octeon_mult_restore3)
-- 
2.25.1

